hphp/runtime/vm/jit/vasm-ppc64.cpp

```diff
@@ -1564,6 +1438,207 @@ void lowerForPPC64(Vunit& unit) {
+ * So we need to patch it to avoid ppc64el unsuported address modes.
+ *
+ * After patching, the Vptr @p will only have either base and index or base and
+ * displacement.
+ */
+void fixVptr(Vout& v, Vptr& p) {
+  // Map all address modes that Vptr can be so it can be handled.
+  enum class AddressModes {
+    Invalid        = 0,
+    Disp           = 1, // Displacement
+    Base           = 2, // Base
+    BaseDisp       = 3, // Base+Displacement
+    Index          = 4, // Index
+    IndexDisp      = 5, // Index+Dispacement
+    IndexBase      = 6, // Index+Base
+    IndexBaseDisp  = 7  // Index+Base+Displacement
+  };
+
+  AddressModes mode = static_cast<AddressModes>(
+                    (((p.disp != 0)     & 0x1) << 0) |
+                    ((p.base.isValid()  & 0x1) << 1) |
+                    ((p.index.isValid() & 0x1) << 2));
+
+  // Index can never be used directly if shifting is necessary. Handling it here
+  uint8_t shift = p.scale == 2 ? 1 :
+                  p.scale == 4 ? 2 :
+                  p.scale == 8 ? 3 : 0;
+
+  if (p.index.isValid() && shift) {
+    Vreg shifted_index_reg = v.makeReg();
+    v << shlqi{shift, p.index, shifted_index_reg, VregSF(RegSF{0})};
+    p.index = shifted_index_reg;
+    p.scale = 1;  // scale is now normalized.
+  }
+
+  // taking care of the displacement, in case it is > 16bits
+  Vreg disp_reg;
+  bool patched_disp = patchImm(Immed(p.disp), v, disp_reg);
+  switch (mode) {
+    case AddressModes::Base:
+    case AddressModes::IndexBase:
+      // ppc64 can handle these address modes. Nothing to do here.
+      break;
+
+    case AddressModes::BaseDisp:
+      // ppc64 can handle this address mode if displacement < 16bits
+      if (patched_disp) {
+        // disp is loaded on a register. Change address mode to kBase_Index
+        p.index = disp_reg;
+        p.disp = 0;
+      }
+      break;
+
+    case AddressModes::Index:
+        // treat it as kBase to avoid a kIndex_Disp asm handling.
+        std::swap(p.base, p.index);
+      break;
+
+    case AddressModes::Disp:
+    case AddressModes::IndexDisp:
+      if (patched_disp) {
+        // disp is loaded on a register. Change address mode to kBase_Index
+        p.base = disp_reg;
+        p.disp = 0;
+      }
+      else {
+        // treat it as kBase_Disp to avoid a kIndex_Disp asm handling.
+        p.base = p.index;
+      }
+      break;
+
+    case AddressModes::IndexBaseDisp: {
+      // This mode is not supported: Displacement will be embedded on Index
+      Vreg index_disp_reg = v.makeReg();
+      lowerImm(Immed(p.disp), v, disp_reg);
+      v << addq{disp_reg, p.index, index_disp_reg, VregSF(RegSF{0})};
+      p.index = index_disp_reg;
+      p.disp = 0;
+      break;
+    }
+
+    case AddressModes::Invalid:
+    default:
+      assert(false && "Invalid address mode");
+      break;
+  }
+}
+
+/* Fallback, when no Vptr is to be fixed */
+template <typename Inst>
+void fixVptrsForPPC64(Vout& v, Inst& inst) {}
```

Comment:
Are there any Vinstrs with Vptr operands that you _don't_ want to be fixed up with this pass?  If not, I think writing a visitor class to use with visitOperands() would be a more robust solution here.